bee-message/src/address/mod.rs

```diff
@@ -67,4 +73,61 @@ impl Address {
         // SAFETY: encoding can't fail as `self` has already been validated and built.
         bech32::encode(hrp.as_ref(), self.pack_to_vec().to_base32(), Variant::Bech32).unwrap()
     }
+
+    ///
+    pub fn unlock(
+        &self,
+        unlock_block: &UnlockBlock,
+        inputs: &[(OutputId, &Output)],
+        context: &mut ValidationContext,
+    ) -> Result<(), ConflictReason> {
+        match (self, unlock_block) {
+            (Address::Ed25519(ed25519_address), UnlockBlock::Signature(unlock_block)) => {
+                if context.unlocked_addresses.contains(self) {
+                    return Err(ConflictReason::SemanticValidationFailed);
+                }
+
+                let Signature::Ed25519(signature) = unlock_block.signature();
+
+                if signature.is_valid(&context.essence_hash, ed25519_address).is_err() {
+                    return Err(ConflictReason::InvalidSignature);
+                }
+
+                context.unlocked_addresses.insert(*self);
+            }
+            (Address::Ed25519(_ed25519_address), UnlockBlock::Reference(_unlock_block)) => {
+                // TODO actually check that it was unlocked by the same signature.
+                if !context.unlocked_addresses.contains(self) {
+                    return Err(ConflictReason::InvalidSignature);
+                }
+            }
+            (Address::Alias(alias_address), UnlockBlock::Alias(unlock_block)) => {
+                // SAFETY: indexing is fine as it is already syntactically verified that indexes reference below.
+                if let (output_id, Output::Alias(alias_output)) = inputs[unlock_block.index() as usize] {
+                    if &alias_output.alias_id().or_from_output_id(output_id) != alias_address.alias_id()
+                        || !context.unlocked_addresses.contains(self)
```

Comment:
I don't really like multiline conditionals (there's a few in other files too), I think it's usually clearer if they're broken down a bit beforehand. At least for conditions that aren't straightforward. That could just be a personal nitpick, though.